### ISSUE TITLE
[BUG] Workspace widget rework (from #189)

### ISSUE THREAD

**Issue by smathles** [Jas-SinghFSU/HyprPanel#190]
**Describe the bug**
Bar after pull #189 still doesn't have parity with waybar functionality. A few notables:
- When on startup, only 1 workspace is shown when I would expect the number of workspaces to be whatever I set the value of `Total Workspaces` to be in dashboard menu (these would be "persistent" workspaces IF the `hide unoccupied` workspaces option is selected). In other words, _I would like workspaces 1, 2, 3, 4 to be shown persistently if I have set my `Total Workspaces` number to 4._ Additional workspaces (>5) should still follow the "hide unoccupied" rules.
##### `Total Workspaces` is set to 4, yet only 2 workspaces shown (only 2 are active):
![2024-08-26T11:50:39,610504185+10:00](https://github.com/user-attachments/assets/010952ed-b784-4c63-a053-5f3f25d965cf)

- When I do have many workspaces open, when I (for example) here have 5 workspaces open with nothing in workspace 3, pressing `$mainMod+shift+3` when on workspace 6, I expect the window to be moved to workspace 3, which will then open inbetween workspaces 2 and 4. However, the behaviour simply keeps it at "index" 5, but "number" 3.
##### Initial state
![2024-08-26T11:39:30,992117013+10:00](https://github.com/user-attachments/assets/92b5f1a8-cb74-49fc-8f56-a4e6ce018199)
##### State after $mainMod+shift+3
![2024-08-26T11:39:39,419665199+10:00](https://github.com/user-attachments/assets/90dd79bb-ce5f-4e68-af7b-b223a13e6894)

- OPTIONAL: As a follow on from this behaviour, when I log into hyprland, only 1 workspace is shown (no. 1). Pressing keybindings to go to workspace 8 will then only show workspace 8 instead of both workspace 1 and 8 (IMPORTANT: This is assuming that workspace 1 is "persistent", which IMO should only be the case only when the `Total Workspaces` is set to 1. Maybe setting this value to zero would the avoid any persistent workspaces again, just like the current behaviour?
![2024-08-26T11:31:22,164784057+10:00](https://github.com/user-attachments/assets/d90f0714-cf63-4db4-aab2-7edb70b6dc1d)
TL;DR: I think that the `Total Workspaces` field should be set to function as `Persistent Workspaces` *if and only if* the `Hide Unoccupied` field is selected.

**Desktop (please complete the following information):**
Arch Linux lastest, hyprland

**Additional Info**
Sorry for the long report oof. In all honesty though this is an awesome improvement in adding waybar's functionality, I appreciate it!

**Comment by Jas-SinghFSU** [comment]
@smathles  I think I have covered all the bases, could you please test all of this in the branch `patch/workspaces`?

**Comment by smathles** [comment]
(Actively testing this branch atm, will report back a little later).

**Comment by Jas-SinghFSU** [comment]
I've gone ahead and covered all the scenarios I think so I'll go ahead and merge it. Our discord conversation basically led me down the right path.

**Comment by Gajus84** [comment]
Thank you for the great work on this!

I believe I found two additional points:

1.) Using things like the pyprland scratchpads will produce non-empty workspaces at -98 for example. Those should be ignored always.

2.) Mask Workspace Numbers on Monitors should also count the unoccupied workspaces. Else it will show wrong number in correspondence to the real workspace number.

**Comment by Jas-SinghFSU** [comment]
@Gajus84 Could you open an issue about 1 please? As for 2, Hyprland doesn't have static workspaces, so there's no concept of empty workspaces occupying space.

For example if you have workspaces:
workspaces: 4 5 6 7 8
mask:            1 2 3 4 5

The mask is easy. But imagine that 6 and 7 were empty. Hyprland only gives us the following information
workspaces: 4 5 8

It doesn't tell us that there is 6 and 7 as well. And we can't guess it either because those 6 and 7 may exist on different monitors (which is funnily enough my exact IRL scenario).

So all we can do is mask what's available:
workspaces: 4 5 8
mask:             1 2 3

**Comment by Gajus84** [comment]
@Jas-SinghFSU  Done for the first one.

As for static workspaces. There is a workaround that I'm using.

You can bind Workspaces to monitors for example like this:


```
workspace=1, monitor:DP-1, default:true
workspace=2, monitor:DP-1, default:true
workspace=3, monitor:DP-1, default:true
workspace=4, monitor:DP-1, default:true
workspace=5, monitor:DP-1, default:true
workspace=6, monitor:DP-1, default:true
workspace=7, monitor:DP-1, default:true
workspace=8, monitor:DP-1, default:true
workspace=9, monitor:DP-1, default:true
workspace=10, monitor:DP-1, default:true

workspace=11, monitor:DP-2, default:true
workspace=12, monitor:DP-2, default:true
workspace=13, monitor:DP-2, default:true
workspace=14, monitor:DP-2, default:true
workspace=15, monitor:DP-2, default:true
workspace=16, monitor:DP-2, default:true
workspace=17, monitor:DP-2, default:true
workspace=18, monitor:DP-2, default:true
workspace=19, monitor:DP-2, default:true
workspace=20, monitor:DP-2, default:true
```

That in combination with a script or a plugin like hypersplit
https://github.com/shezdy/hyprsplit
will allow me to use Hyprland like AwesomeWM/DWM with fixed numbered workspaces per monitor. 
Waybar allows for custom icons so I used this to "bind" the numbers. Like 11 displayed as 1, 12 as 2 etc.
Basically the same thing as HyprBar does for Window Titels if I'm correct? Maybe this could be ported to workspaces? Should I open a feature request for that?
